pkg/engine/internal/executor/project.go

```diff
@@ -11,43 +11,90 @@ import (
 	"github.com/grafana/loki/v3/pkg/engine/internal/types"
 )
 
-func NewProjectPipeline(input Pipeline, columns []physical.ColumnExpression, evaluator *expressionEvaluator) (*GenericPipeline, error) {
-	// Get the column names from the projection expressions
-	columnNames := make([]string, len(columns))
-
-	for i, col := range columns {
-		if colExpr, ok := col.(*physical.ColumnExpr); ok {
-			columnNames[i] = colExpr.Ref.Column
-		} else {
-			return nil, fmt.Errorf("projection column %d is not a column expression", i)
-		}
-	}
-
+func NewProjectPipeline(input Pipeline, projection *physical.Projection, evaluator *expressionEvaluator) (*GenericPipeline, error) {
 	return newGenericPipeline(Local, func(ctx context.Context, inputs []Pipeline) state {
 		// Pull the next item from the input pipeline
 		input := inputs[0]
 		batch, err := input.Read(ctx)
 		if err != nil {
 			return failureState(err)
 		}
+		defer batch.Release()
+
+		columns := projection.Expressions
+		// short circuit if there are no columns to project, treat as a select *
+		if len(columns) == 0 {
+			projectedRecord := array.NewRecord(batch.Schema(), batch.Columns(), batch.NumRows())
+			return successState(projectedRecord)
+		}
 
-		projected := make([]arrow.Array, 0, len(columns))
-		fields := make([]arrow.Field, 0, len(columns))
+		columnOrder := []string{}
+		projected := map[string]arrow.Array{}
+		fields := map[string]arrow.Field{}
 
-		for i := range columns {
-			vec, err := evaluator.eval(columns[i], batch)
+		for _, col := range columns {
+			vec, err := evaluator.eval(col, batch)
 			if err != nil {
 				return failureState(err)
 			}
-			fields = append(fields, arrow.Field{Name: columnNames[i], Type: vec.Type().ArrowType(), Metadata: types.ColumnMetadata(vec.ColumnType(), vec.Type())})
-			projected = append(projected, vec.ToArray())
+
+			switch col := col.(type) {
+			case *physical.ColumnExpr:
+				columnName := col.Ref.Column
+				columnOrder = append(columnOrder, columnName)
+				fields[columnName] = arrow.Field{
+					Name:     columnName,
+					Type:     vec.Type().ArrowType(),
+					Metadata: types.ColumnMetadata(vec.ColumnType(), vec.Type()),
+				}
+				projected[columnName] = vec.ToArray()
+			case *physical.UnaryExpr:
+				if arrStruct, ok := vec.ToArray().(*array.Struct); ok {
+					defer arrStruct.Release()
+					for i := range arrStruct.NumField() {
+						arr := arrStruct.Field(i)
+
+						structSchema, ok := arrStruct.DataType().(*arrow.StructType)
+						if !ok {
+							return failureState(fmt.Errorf("unexpected type for struct field %d, got %T", i, arrStruct.DataType()))
+						}
+						field := structSchema.Field(i)
+
+						if _, ok := fields[field.Fingerprint()]; ok {
+							continue
+						}
+
+						columnOrder = append(columnOrder, field.Fingerprint())
+						fields[field.Fingerprint()] = field
+						projected[field.Fingerprint()] = arr
```

Comment:
You can use `field.Name()` instead of `field.Fingerprint()`, since the name is the fully qualified column name.